model/task_lifecycle.go

```diff
@@ -506,6 +506,49 @@ func MarkEnd(t *task.Task, caller string, finishTime time.Time, detail *apimodel
 	return nil
 }
 
+// UpdateBlockedDependencies traverses the dependency graph and recursively sets each
+// parent dependency as unattainable in depending tasks.
+func UpdateBlockedDependencies(t *task.Task) error {
+	dependentTasks, err := t.FindAllUnmarkedBlockedDependencies()
+	if err != nil {
+		return errors.Wrapf(err, "can't get tasks depending on task '%s'", t.Id)
+	}
+
+	for _, dependentTask := range dependentTasks {
+		if err = dependentTask.MarkUnattainableDependency(t, true); err != nil {
+			return errors.Wrap(err, "error marking dependency unattainable")
+		}
+		if err = build.SetCachedTaskBlocked(dependentTask.BuildId, dependentTask.Id, true); err != nil {
+			return errors.Wrap(err, "error marking cached task as blocked")
+		}
+		if err = UpdateBlockedDependencies(&dependentTask); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+	return nil
+}
+
+func UpdateUnblockedDependencies(t *task.Task) error {
+	blockedTasks, err := t.FindAllMarkedUnattainableDependencies()
+	if err != nil {
+		return errors.Wrap(err, "can't get dependencies marked unattainable")
+	}
+
+	for _, blockedTask := range blockedTasks {
+		if err = blockedTask.MarkUnattainableDependency(t, false); err != nil {
+			return errors.Wrap(err, "error marking dependency attainable")
+		}
+		if err = build.SetCachedTaskBlocked(blockedTask.BuildId, blockedTask.Id, false); err != nil {
```

Comment:
it will only be unblocked if **all** it's unattainable fields are false

Reply:
Interesting observation. Perhaps instead of explicitly passing in True and False we should pass in dependentTask.Blocked() and let that logic handle itself.